### DISCUSSION TITLE
Code and Project Structure

### DISCUSSION THREAD

**Discussion by kespinola** [holaplex/holaplex#343] [Ideas]
## Goal

Document code and project structure for Holaplex client applications.

### (current): Many components
The code is broken down into mostly self contained components even if they are not reused in other places. Data fetching is often done on a pr component level. The new Feed as an example is structured as
* /page/Feed
    * FeedContainer
        *  FollowingFeed
        * FeedItem
    * WhoToFollow
        * FollowListItem
    * ProfilePopover
    * MyActivityList
        * ActivityCard
* pros
    * IMO code is easer to reason about
    * can more easily be moved around
    * easy to send someone to fix/build in one file
    * relatively short files
* cons
    * some prop drilling

### Few components
Tries to do all kinds of data fetching in the actual page component and keep most of the HTML in that file as well. The feed structure would be something like
* /page/feed
    * FeedItem(maybe)
    * ProfilePopover
* pros
    * centralized data fetching

**Comment by austinmilt** [comment]
My two-cents on code structure, in no particular order, and mostly stuff we've all heard before. I think the discussion is a bit more amorphous than "many vs few components," as the "right answer" to the question "should this be another component/class/method/etc?" is usually pretty situational.

- put "stuff" close to its associations rather than close to similar types, e.g. `ComponentAProps` goes close to `ComponentA` rather than `ComponentBProps`. This applies to variables, methods, classes, components, files, folders, modules, etc.
- "stuff" should be narrowly-scoped to the point that it can be clearly and succinctly named, e.g. a method that asks for payment confirmation, sends money, and then displays the result would ideally be composed of three separate method calls, with an outer method named like `submitOrder`, rather than expanding those actions within the method. Any method more than 20-30 lines of code should probably be split into multiple methods. Any component more than a couple hundred lines should probably be split into multiple components. This applies to methods, classes, components, files, folders, etc.
- code should be shallowly nested. If your render method nests elements 10 levels deep, it probably needs to be broken into multiple components. If your method has more than 4 nested control blocks (if/for/while/try/etc) then it probably needs to be broken into multiple methods.
- avoid mixing markup (~html) and logic while taking advantage of react. `render` should primarily show how your document is structured, not where/how the data come from or are processed. If your inline anonymous method for conditionally rendering an (series of) elements is more than comfortably fits on one line, put it in a method or extract it to a component.
- specifically about tailwind from a tailwind noob: Tailwind utility classes show how/what an element does/is, but not why. We should have a way to identify the purpose of an element even when that element isnt its own react component. Something like an `id` attribute, e.g. `left-panel-container`. Even just a line comment like `//left-panel-container`.

I dont think any of this necessarily dictates how many components we have or the degree to which we are centralizing data fetching, though it definitely biases toward more components/files, and more smaller network requests.

**Comment by kespinola** [comment]
## Holaplex Next App Project Structure

```
/pages
  _app.tsx
  index.tsx
  feed.tsx
  /profiles
    [address].tsx
/src
  /components # UI building blocks. Shared between views.
    /Button
      Button.tsx
      index.ts
    /List
      List.tsx
      index.ts
    /Card
      Card.tsx
      index.ts
  /hooks # bundle handler and helper functions for page views in hooks
  /layouts # shared page layouts for achieving nested views
  /contexts # react contexts 
  /providers # react providers (primarily exposed in the _app.tsx or layouts)
  /modules # internal sdk (the start of npm packages. collection of like objects and functions)
     /addresses
       index.ts
    /ipfs
      index.ts
      client.ts
    /arweave
      index.ts
      client.ts
  /queries # graphql queries available for use in views
```

## Code Philosophies 
- All markup (whether native or UI components) is placed in the next page of the route. Pages in most cased should only have 2-levels of component nesting.
- Components should be formed from native elements or re-export existing component packages. They may reference sibling component files.
- Hooks should package handler callbacks, useEffects, and queries leveraged by the page. Hooks are what keep the page from becoming bloated.
- Any query needed for a page should be done within the next page. This allows for a single loading experience even if page formed from many queries.
- _app.tsx will setup providers and the layout.
- The default layout is the header and footer wrapping the page provided content.
- When nested views are required use a layout.

**Reply by kespinola** [reply]
Looking for this to be our default layout.

![Default Layout](https://user-images.githubusercontent.com/2388118/164489849-b71ad945-a9c0-4264-b150-db1455837d56.png)

https://nextjs.org/docs/basic-features/layouts

**Reply by kevinrodriguez-io** [reply]
I would like us to move the layout from an "App" Level perspective into a page basis, as the first wrapper thing, this will ease the usage of different layouts in different contexts (For example, you don't always want to show the same header bar with an omni search if you're on home vs if you're on a feed page)

**Comment by kristianeboe** [comment]
[Still in draft! Just need a place to save it 😅  ]
Here is my ideal project structure

```
/pages
  _app.tsx
  index.tsx
  about.tsx
  // other "landing pages"
  /feed
      index.tsx
      discovery.tsx // a different feed tab, not based on who you follow
      who-to-follow.tsx // a separate page for who to follow
  /discover/[discoveryTab].tsx
  /profiles
    [address].tsx
         collected
         created
         offers
         activity
        (or a catch all [profileTab].tsx)
/src
  /components # UI building blocks. Shared between views.
    /shared/
          Button.tsx
          Modal.tsx
          Footer.tsx
          Nav.tsx
    /Feed
        FeedContainer
        FollowingFeed
        FeedItem
        WhoToFollow
        FollowListItem
        ProfilePopover
    /Profile
        NFTGrid
        NFTCard
        ActivityList
        ActivityCard
    /Home
        /FeaturedMarketplacesSection
        /FeaturedStorefrontsSection
        /FeaturedFollowersSection
        /HeroSection
        /MarketplacePreview
        /ListingPreview
  /hooks # bundle handler and helper functions for page views in hooks
  /layouts # shared page layouts for achieving nested views
      LandingLayout // for home page and other landing pages
      ProfileLayout
      FeedLayout (possibly)
      
  /contexts # react contexts 
  /providers # react providers (primarily exposed in the _app.tsx or layouts)
  /modules # internal sdk (the start of npm packages. collection of like objects and functions)
     /addresses
       index.ts
    /ipfs
      index.ts
      client.ts
    /arweave
      index.ts
      client.ts
  /queries # graphql queries available for use in views
```

Key takeaways / differences
* quite a few more feature/domain folders to organize components even if they are not used extensively or at all in the rest of the project.
* components export directly from their own file, instead of from an index.ts file (I.E. /Button.tsx instead of /Button/index.ts)
* contexts and providers merged into one folder
* leave data fetching logic to the component that actually needs the data, or hoist it if it makes sense to do so. Rely on cached queries to avoid repeated api calls while navigating around.

**Reply by kespinola** [reply]
```
    /shared/
          Button.tsx
          Modal.tsx
          Footer.tsx
          Nav.tsx
    /Feed
        FeedContainer
        FollowingFeed
        FeedItem
        WhoToFollow
        FollowListItem
        ProfilePopover
    /Profile
        NFTGrid
        NFTCard
        ActivityList
        ActivityCard
    /Home
        /FeaturedMarketplacesSection
        /FeaturedStorefrontsSection
        /FeaturedFollowersSection
        /HeroSection
        /MarketplacePreview
        /ListingPreview
```

Whenever you start nesting its difficult to establish a clear pattern. This may be personal experience but never have luck trying to group things. Flatter folder structures always seem to work out better for me. 

And again I don't love the page specific components. If you have a good base component library like antd (but not antd) you don't need to have this extra grouping. These are views that should be looping over data and displaying it. I don't like over complicated the view layer. Pages from nextjs are enough.

I am fine with not doing the extra folder per component with the index.ts ie 

```
src/
  components/
    Button.tsx
    List.tsx
```

**Reply by kespinola** [reply]
@kristianeboe if you want to get consensus from the discovery team. I'll follow whatever you all like best.
